File: src/jobs/process_thread.rs
use crate::config::CONFIG;
use anyhow::Result;
use std::sync::Arc;
use twilight_http::Client;
use twilight_model::channel::Channel;

pub async fn process_thread(client: Arc<Client>, channel: Channel) -> Result<()> {
    if Some(CONFIG.help_channel_id) != channel.parent_id {
        return Ok(());
    }

    let mut applied_tags = channel.applied_tags.clone().unwrap_or_default();
    let has_solved = applied_tags.iter().any(|&v| v == CONFIG.solved_tag_id);
    let has_unsolved = applied_tags.iter().any(|&v| v == CONFIG.unsolved_tag_id);

    let thread_name = channel.name.clone().unwrap_or_default();

    if has_solved && has_unsolved {
        println!("Removing unsolved tag from existing thread: \"{thread_name}\"");
        applied_tags.retain(|&v| v != CONFIG.unsolved_tag_id);
    } else if !has_solved && !has_unsolved {
        println!("Adding unsolved tag to existing thread: \"{thread_name}\"");
        applied_tags.push(CONFIG.unsolved_tag_id);
    } else {
        return Ok(());
    }

    client
        .update_thread(channel.id)
        .applied_tags(Some(&applied_tags))
        .await?;

    Ok(())
}

File: src/jobs/processThread.ts
import Discord from "discord.js";
import { HELP_CHANNEL_ID, SOLVED_TAG_ID, UNSOLVED_TAG_ID } from "../constants";
import log from "@osyris/log";

export async function processThread(channel: Discord.AnyThreadChannel) {
	if (HELP_CHANNEL_ID !== channel.parentId) return;

	const hasSolved = channel.appliedTags.includes(SOLVED_TAG_ID);
	const hasUnsolved = channel.appliedTags.includes(UNSOLVED_TAG_ID);

	const threadName = channel.name;

	const { ownerId, appliedTags } = channel;
	const metadata = { threadName, ownerId, appliedTags };

	if (hasSolved && hasUnsolved) {
		log.info(`Removing unsolved tag from existing thread: "${threadName}"`, metadata);
		await channel.setAppliedTags(appliedTags.filter(tag => tag !== UNSOLVED_TAG_ID));
	} else if (!hasSolved && !hasUnsolved) {
		log.info(`Adding unsolved tag to existing thread: "${threadName}"`, metadata);
		await channel.setAppliedTags([...appliedTags, UNSOLVED_TAG_ID]);
	}
}
